lisa/sut_orchestrator/baremetal/platform_.py

```diff
@@ -66,7 +67,49 @@ def _prepare_environment(self, environment: Environment, log: Logger) -> bool:
         return self._configure_node_capabilities(environment, log, client_capabilities)
 
     def _deploy_environment(self, environment: Environment, log: Logger) -> None:
-        # copy build (shared, check if it's copied)
+        # process the cluster elements from runbook
+        self._process_cluster_elements(environment, log)
+
+        # deploy cluster
+        self.cluster.deploy(environment)
+
+        if self._cluster_runbook.ready_checker:
+            ready_checker = self.ready_checker_factory.create_by_runbook(
+                self._cluster_runbook.ready_checker
+            )
+
+        for index, node in enumerate(environment.nodes.list()):
+            node_context = get_node_context(node)
+
+            # ready checker
+            if ready_checker:
+                ready_checker.is_ready(node)
+
+            # get ip address
+            if self._cluster_runbook.ip_getter:
+                ip_getter = self.ip_getter_factory.create_by_runbook(
+                    self._cluster_runbook.ip_getter
+                )
+                node_context.connection.address = ip_getter.get_ip()
+
+            assert isinstance(node, RemoteNode), f"actual: {type(node)}"
+            node.name = f"node_{index}"
+            try_connect(node_context.connection)
+
+        self._log.debug(f"deploy environment {environment.name} successfully")
+
+    def copy(self, build_schema: BuildSchema, sources_path: List[Path]) -> None:
+        if sources_path:
+            build = self.build_factory.create_by_runbook(build_schema)
+            build.copy(
+                sources_path=sources_path,
+                files_map=build_schema.files,
+            )
+        else:
+            self._log.debug("no copied source path specified, skip copy")
+
+    def _process_cluster_elements(self, environment: Environment, log: Logger) -> None:
```

Comment:
Call it like `_predeploy_environment`. It should be clearer than "process".

Reply:
the change is done, thanks